### DISCUSSION TITLE
Feature request: Custom error messages for Type.Union()

### DISCUSSION THREAD

**Discussion by EvHaus** [sinclairzx81/typebox#936] [Ideas]
It would be nice if `Type.Union()` supported the `errorMessage` option like other types. This would allow for something more user-friendly, like:

```ts
Type.Union([Type.Literal('Canada'), Type.Literal('USA')], {
   errorMessage: "Only North American countries are allowed"
})
```

Instead of the current generic "Expected union value" message that is returned.

At the moment I'm working around this with a custom `SetErrorFunction` definition.

I tried using the `Type.Unsafe` solution recommended [here](https://github.com/sinclairzx81/typebox/issues/681#issuecomment-1836878768) but it didn't work for me, and wouldn't `Type.Unsafe` also cause the type to be `unknown`?

**Comment by sinclairzx81** [comment]
@EvHaus Hi,

TypeBox doesn't actually have a concept of `errorMessage` at a schema level; it only has a [ValueErrorType](https://github.com/sinclairzx81/typebox/blob/master/src/errors/errors.ts#L95-L160) (which includes a defacto us-EN message associated which is emitted on a ValueError). This reason for this is that `errorMessage` is not actually a known Json Schema keyword, and TypeBox tries to avoid associating any non-standard keywords with downstream logic (such as error handling).

This said, TypeBox doesn't prohibit end users from defining an `errorMessage` on options for any schema (as some Json Schema validators understand this keyword), but as TypeBox doesn't associate logic with unknown keywords, the recommendation is to use the SetErrorFunction to handle custom message generation.

The reference issue you linked was from a while back, here is an updated version.

```typescript
import { SetErrorFunction, DefaultErrorFunction } from '@sinclair/typebox/errors'
import { Value } from '@sinclair/typebox/value'
import { Type } from '@sinclair/typebox'

SetErrorFunction((parameter) => parameter.schema.errorMessage ?? DefaultErrorFunction(parameter))

const T = Type.Union([
  Type.Literal('A'),
  Type.Literal('B')
], {
  errorMessage: 'Expected either A or B'
})

console.log(Value.Errors(T, 'C').First())  // Expected either A or B
```
Hope this helps
S

**Reply by EvHaus** [reply]
> TypeBox doesn't actually have a concept of errorMessage at a schema level;

Gah, that's my bad. It's actually [an Elysia thing](https://elysiajs.com/validation/primitive-type.html#custom-error) and not a Typebox thing. This is the con of having other libraries extend Typebox -- it's not always super clear where the problem is.

Thanks for your help again!
